File: TextosJapones/src/main/java/org/jisho/textosJapones/util/configuration/Configuracao.java
package org.jisho.textosJapones.util.configuration;

import org.jisho.textosJapones.components.notification.Alertas;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.util.Properties;

public class Configuracao {

    private static final Logger LOGGER = LoggerFactory.getLogger(Configuracao.class);

    private static Properties props = null;
    private static Properties secrets = null;

    public static void saveProperties(String winrar, String commictagger) {
        try (OutputStream os = new FileOutputStream("db.properties")) {
            if (props.containsKey("caminho_winrar"))
                props.replace("caminho_winrar", winrar);
            else
                props.put("caminho_winrar", winrar);

            if (props.containsKey("caminho_commictagger"))
                props.replace("caminho_commictagger", commictagger);
            else
                props.put("caminho_commictagger", commictagger);
            props.store(os, "");
        } catch (IOException e) {
            Alertas.Tela_Alerta("Erro ao salvar o properties", e.getMessage());
            
            LOGGER.error(e.getMessage(), e);
        }
    }

    public static void saveProperties(String server, String port, String user, String pws, String mysql, String base) {
        try (OutputStream os = new FileOutputStream("db.properties")) {
            if (props.containsKey("server"))
                props.replace("server", server);
            else
                props.put("server", server);

            if (props.containsKey("port"))
                props.replace("port", port);
            else
                props.put("port", port);

            if (props.containsKey("user"))
                props.replace("user", user);
            else
                props.put("user", user);

            if (props.containsKey("password"))
                props.replace("password", pws);
            else
                props.put("password", pws);

            if (props.containsKey("base"))
                props.replace("base", base);
            else
                props.put("base", base);

            if (props.containsKey("caminho_mysql"))
                props.replace("caminho_mysql", mysql);
            else
                props.put("caminho_mysql", mysql);

            props.store(os, "");
        } catch (IOException e) {
            Alertas.Tela_Alerta("Erro ao salvar o properties", e.getMessage());
            
            LOGGER.error(e.getMessage(), e);
        }
    }

    private static void createProperties() {
        props = new Properties();
        try (OutputStream os = new FileOutputStream("db.properties")) {
            props.clear();
            props.setProperty("server", "");
            props.setProperty("port", "");
            props.setProperty("user", "");
            props.setProperty("password", "");
            props.setProperty("base", "");
            props.setProperty("caminho_mysql", "");
            props.setProperty("caminho_winrar", "");
            props.setProperty("caminho_commictagger", "");
            props.store(os, "");
        } catch (IOException e) {
            Alertas.Tela_Alerta("Erro ao salvar o properties", e.getMessage());
            
            LOGGER.error(e.getMessage(), e);
        }
    }

    private static Properties loadProperties() {
        File f = new File("db.properties");
        if (!f.exists())
            createProperties();

        if (props == null)
            props = new Properties();

        try (FileInputStream fs = new FileInputStream("db.properties")) {
            props.load(fs);
            return props;
        } catch (IOException e) {
            Alertas.Tela_Alerta("Erro ao carregar o properties", e.getMessage());
            
            LOGGER.error(e.getMessage(), e);
        }
        return null;
    }

    public static Properties getProperties() {
        if (props == null)
            loadProperties();
        return props;
    }

    public static String getCaminhoWinrar() {
        return getProperties().getProperty("caminho_winrar");
    }

    // -------------------------------------------------------------------------------------------------

    private static Properties loadSecrets() {
        File f = new File("secrets.properties");

        if (!f.exists()) {
            try (OutputStream os = new FileOutputStream("secrets.properties")) {
                Properties props = new Properties();

                props.setProperty("my_anime_list_client_id", "");
                props.store(os, "");
            } catch (IOException e) {
                Alertas.Tela_Alerta("Erro ao salvar o secrets", e.getMessage());
                
                LOGGER.error(e.getMessage(), e);
            }
        }

        try (FileInputStream fs = new FileInputStream("secrets.properties")) {
            Properties props = new Properties();
            props.load(fs);
            return props;
        } catch (IOException e) {
            Alertas.Tela_Alerta("Erro ao carregar o secrets", e.getMessage());
            
            LOGGER.error(e.getMessage(), e);
        }
        return null;
    }

    public static Properties getSecrets() {
        if (secrets == null)
            secrets = loadSecrets();
        return secrets;
    }

    public static String getMyAnimeListClient() {
        return getSecrets().getProperty("my_anime_list_client_id");
    }

}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/util/configuration/Configuracao.kt
package br.com.fenix.processatexto.util.configuration

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.util.*


object Configuracao {

    private val LOGGER: Logger = LoggerFactory.getLogger(Configuracao::class.java)

    private val secrets: Properties = Properties()
    private val properties: Properties = Properties()

    init {
        loadProperties()
        loadSecrets()
    }

    fun saveProperties() {
        try {
            FileOutputStream("app.properties").use { os ->
                properties.store(os, "")
            }
        } catch (e: IOException) {
            //Alertas.Tela_Alerta("Erro ao salvar o properties", e.message )
            LOGGER.error(e.message, e)
        }
    }

    private fun createProperties() {
        try {
            FileOutputStream("app.properties").use { os ->
                properties.clear()
                properties.setProperty("connection.server", "")
                properties.setProperty("connection.port", "")
                properties.setProperty("connection.user", "")
                properties.setProperty("connection.password", "")
                properties.setProperty("connection.database", "")
                properties.setProperty("caminho.mysql", "")
                properties.setProperty("caminho.winrar", "")
                properties.setProperty("caminho.commictagger", "")
                properties.setProperty("caminho.legenda", "")
                properties.store(os, "")
            }
        } catch (e: IOException) {
            //Alertas.Tela_Alerta("Erro ao salvar o properties", e.message )
            LOGGER.error(e.message, e)
        }
    }

    private fun loadProperties(): Properties {
        val f = File("app.properties")
        if (!f.exists())
            createProperties()
        try {
            FileInputStream("app.properties")
                .use { fs ->
                    properties.load(fs)
                    return properties
                }
        } catch (e: IOException) {
            //Alertas.Tela_Alerta("Erro ao carregar o properties", e.message )
            LOGGER.error(e.message, e)
            throw Exception("Erro ao carregar o properties")
        }
    }

    var server: String = ""
        set(value) {
            properties["connection.server"] = value
            field = value
        }
        get() = properties.getProperty("connection.server", "")

    var port: String = ""
        set(value) {
            properties["connection.port"] = value
            field = value
        }
        get() = properties.getProperty("connection.port", "")

    var user: String = ""
        set(value) {
            properties["connection.user"] = value
            field = value
        }
        get() = properties.getProperty("connection.user", "")

    var password: String = ""
        set(value) {
            properties["connection.password"] = value
            field = value
        }
        get() = properties.getProperty("connection.password", "")

    var database: String = ""
        set(value) {
            properties["connection.database"] = value
            field = value
        }
        get() = properties.getProperty("connection.database", "")

    var caminhoMysql: String = ""
        set(value) {
            properties["caminho.mysql"] = value
            field = value
        }
        get() = properties.getProperty("caminho.mysql", "")

    var caminhoCommicTagger: String = ""
        set(value) {
            properties["caminho.commictagger"] = value
            field = value
        }
        get() = properties.getProperty("caminho.commictagger", "")

    var caminhoLegenda: String = ""
        set(value) {
            properties["caminho.legenda"] = value
            field = value
        }
        get() = properties.getProperty("caminho.legenda", "")

    var caminhoWinrar: String = ""
        set(value) {
            properties["caminho.winrar"] = value
            field = value
        }
        get() = properties.getProperty("caminho.winrar", "")

    var caminhoSalvoArquivo: String = ""
        set(value) {
            properties["caminho.ultimo_salvo_arquivo"] = value
            field = value
            saveProperties()
        }
        get() = properties.getProperty("caminho.ultimo_salvo_arquivo", "")


    // -------------------------------------------------------------------------------------------------
    private fun loadSecrets(): Properties {
        val f = File("secrets.properties")
        if (!f.exists()) {
            try {
                FileOutputStream("secrets.properties").use { os ->
                    secrets.setProperty("my_anime_list_client_id", "")
                    secrets.store(os, "")
                }
            } catch (e: IOException) {
                //Alertas.Tela_Alerta("Erro ao salvar o secrets", e.message )
                LOGGER.error(e.message, e)
            }
        }
        try {
            FileInputStream("secrets.properties").use { fs ->
                secrets.load(fs)
                return secrets
            }
        } catch (e: IOException) {
            //Alertas.Tela_Alerta("Erro ao carregar o secrets", e.message )
            LOGGER.error(e.message, e)
            throw Exception("Erro ao carregar o secrets")
        }
    }

    val myAnimeListClient: String get() = secrets.getProperty("my_anime_list_client_id", "")
}